core/src/main/scala/kanaloa/reactive/dispatcher/metrics/MetricsCollector.scala

```diff
@@ -1,24 +1,199 @@
 package kanaloa.reactive.dispatcher.metrics
 
-import akka.actor.ActorSystem
-import com.typesafe.config.{ConfigFactory, Config}
+import java.time.{LocalDateTime ⇒ Time}
 
-trait MetricsCollector {
-  def send(metric: Metric): Unit
-}
+import akka.actor._
+import kanaloa.reactive.dispatcher.metrics.Metric._
+import kanaloa.reactive.dispatcher.metrics.MetricsCollector._
+
+import scala.concurrent.duration._
+import kanaloa.util.FiniteCollection._
+import kanaloa.util.Java8TimeExtensions._
+import akka.agent.Agent
+
+/**
+ * A metrics collector to which all [[Metric]] are sent to.
+  *  It forwards incoming [[Metric]]s to [[reporter]].
+  *  It collects performance [[Sample]] from [[WorkCompleted]] and [[WorkFailed]]
+  *  when the system is in fullyUtilized state, namely when number of idle workers is less than [[MetricsCollectorSettings]]
+  *  It can be subscribed using [[Subscribe]] message.
+  *  It publishes [[Sample]]s collected to subscribers.\
+  *  It also publishes [[PartialUtilization]] number to subscribers.
+ * @param reporter
+ * @param settings
+ */
+class MetricsCollector(
+  reporter: Option[Reporter],
+  settings: MetricsCollectorSettings
+
+) extends Actor {
+  import settings._
+  var subscribers: Set[ActorRef] = Set.empty
+
+  val scheduledSampling = {
+    import context.dispatcher
+    context.system.scheduler.schedule(
+      sampleRate,
+      sampleRate,
+      self,
+      AddSample
+    )
+  }
+
+  override def postStop(): Unit = {
+    scheduledSampling.cancel()
+    super.postStop()
+  }
+
+  def receive = partialUtilized(None)
+
+  def handleSubscriptions: Receive = {
+    case Subscribe(s) ⇒
+      subscribers += s
+      context watch s
+    case Unsubscribe(s) ⇒
+      subscribers -= s
+      context unwatch s
+    case Terminated(s) ⇒
+      subscribers -= s
+  }
+
+  def publishUtilization(idle: Int, poolSizeO: Option[Int]): Unit = {
```

Comment:
Again this publication is internal, not public API. So I prefer sticking to whatever is needed rather than what could be more informative, we could always move to provide more information when we need so right? WDTY?